tests/test_type_caster_bare_interface.cpp

```diff
@@ -0,0 +1,215 @@
+mpty const& rtrn_cref() { static mpty obj; return obj; }
+mpty&       rtrn_mref() { static mpty obj; return obj; }
+mpty const* rtrn_cptr() { return new mpty; }
+mpty*       rtrn_mptr() { return new mpty; }
+
+const char* pass_valu(mpty)        { return "load_valu"; }
+const char* pass_rref(mpty&&)      { return "load_rref"; }
+const char* pass_cref(mpty const&) { return "load_cref"; }
+const char* pass_mref(mpty&)       { return "load_mref"; }
+const char* pass_cptr(mpty const*) { return "load_cptr"; }
+const char* pass_mptr(mpty*)       { return "load_mptr"; }
+
+std::shared_ptr<mpty>       rtrn_shmp() { return std::shared_ptr<mpty      >(new mpty); }
+std::shared_ptr<mpty const> rtrn_shcp() { return std::shared_ptr<mpty const>(new mpty); }
+
+const char* pass_shmp(std::shared_ptr<mpty>)       { return "load_shmp"; }
+const char* pass_shcp(std::shared_ptr<mpty const>) { return "load_shcp"; }
+
+std::unique_ptr<mpty>       rtrn_uqmp() { return std::unique_ptr<mpty      >(new mpty); }
+std::unique_ptr<mpty const> rtrn_uqcp() { return std::unique_ptr<mpty const>(new mpty); }
+
+const char* pass_uqmp(std::unique_ptr<mpty>)       { return "load_uqmp"; }
+const char* pass_uqcp(std::unique_ptr<mpty const>) { return "load_uqcp"; }
+
+// clang-format on
+
+} // namespace type_caster_bare_interface
+} // namespace pybind11_tests
+
+namespace pybind11 {
+namespace detail {
+
+using namespace pybind11_tests::type_caster_bare_interface;
+
+template <>
+struct type_caster<mpty> {
+    static constexpr auto name = _<mpty>();
+
+    // static handle cast(mpty, ...)
+    // is redundant (leads to ambiguous overloads).
+
+    static handle cast(mpty && /*src*/, return_value_policy /*policy*/, handle /*parent*/) {
+        return str("cast_rref").release();
+    }
+
+    static handle cast(mpty const & /*src*/, return_value_policy /*policy*/, handle /*parent*/) {
+        return str("cast_cref").release();
+    }
+
+    static handle cast(mpty & /*src*/, return_value_policy /*policy*/, handle /*parent*/) {
+        return str("cast_mref").release();
+    }
+
+    static handle cast(mpty const *src, return_value_policy /*policy*/, handle /*parent*/) {
+        delete src;
+        return str("cast_cptr").release();
+    }
+
+    static handle cast(mpty *src, return_value_policy /*policy*/, handle /*parent*/) {
+        delete src;
+        return str("cast_mptr").release();
+    }
+
+    template <typename T_>
+    using cast_op_type = conditional_t<
+        std::is_same<remove_reference_t<T_>, mpty const *>::value,
+        mpty const *,
+        conditional_t<
+            std::is_same<remove_reference_t<T_>, mpty *>::value,
+            mpty *,
+            conditional_t<
+                std::is_same<T_, mpty const &>::value,
+                mpty const &,
+                conditional_t<std::is_same<T_, mpty &>::value,
+                              mpty &,
+                              conditional_t<std::is_same<T_, mpty &&>::value, mpty &&, mpty>>>>>;
```

Comment:
What about going `else if` style on this, to avoid indentation? (similar to [test_builtin_casters.cpp](https://github.com/pybind/pybind11/blob/0432ae7c5222ecb5d6e81babf996d3777bd94bf3/tests/test_builtin_casters.cpp#L42))

```suggestion
    using cast_op_type =
        conditional_t<std::is_same<remove_reference_t<T_>, mpty const *>::value,
            mpty const *,
        conditional_t<std::is_same<remove_reference_t<T_>, mpty *>::value,
            mpty *,
        conditional_t<std::is_same<T_, mpty const &>::value,
            mpty const &,
        conditional_t<std::is_same<T_, mpty &>::value,
            mpty &,
        conditional_t<std::is_same<T_, mpty &&>::value,
            mpty &&,
            /* else */ mpty>>>>>;
```